03-task-manager/starter/controllers/tasks.js

```diff
@@ -1,30 +1,127 @@
 const Task = require("../models/task");
+const asyncWrapper = require("../middleware/async");
+const { createCustomError } = require("../errors/custom-error");
 
-const getAllTasks = (req, res)=>{
-    res.send("get all tasks")
-}
+const getAllTasks = asyncWrapper(async (req, res) => {
+  //   try {
+  const tasks = await Task.find({});
 
-const createTask = async (req, res)=>{
-    const task = await Task.create(req.body)
-    res.status(200).json({task})
-}
+  // res.status(200).json({ tasks });
+  //equal
+  // res.status(200).json({ tasks: tasks });
+  // res.status(200).json({ tasks, amount: tasks.length });
+  // res
+  //   .status(200)
+  //   .json({ success: true, data: { tasks, nbHits: tasks.length } });
+  res
+    .status(200)
+    .json({ status: "success", data: { tasks, nbHits: tasks.length } });
+  //   } catch (error) {
+  //     res.status(500).json({ msg: error });
+  //   }
+});
 
-const getTask = (req, res)=>{
-    res.json({id: req.params.id})
-}
+const createTask = asyncWrapper(async (req, res) => {
+  //   try {
+  const task = await Task.create(req.body);
+  res.status(200).json({ task });
+  //   } catch (error) {
+  //     res.status(500).json({ msg: error });
+  //   }
+});
 
-const updateTask = (req, res)=>{
-    res.send("update task")
-}
+const getTask = asyncWrapper(async (req, res, next) => {
+  //   res.json({ id: req.params.id });
+  //   try {
+  const { id: taskID } = req.params;
+  const task = await Task.findOne({ _id: taskID });
+  if (!task) {
+    return next(createCustomError(`no task was found with id: ${taskID}`, 404));
+    // const error = new Error("Not found");
+    // error.status = 404;
+    // return next(error);
+    // return res
+    //   .status(404)
+    //   .json({ msg: `no task was found with id: ${taskID}` });
+  }
+  res.status(200).json({ task });
+  //   } catch (error) {
+  //     res.status(500).json({ msg: error });
+  //   }
+});
 
-const deleteTask = (req, res)=>{
-    res.send("delete task")
-}
+const deleteTask = asyncWrapper(async (req, res) => {
+  //   res.send("delete task");
+  //   try {
+  const { id: taskID } = req.params;
+  const task = await Task.findOneAndDelete({ _id: taskID });
+  if (!task) {
+    return next(createCustomError(`no task was found with id: ${taskID}`, 404));
+    // return res
+    //   .status(404)
+    //   .json({ msg: `no task was found with id: ${taskID}` });
+  }
+  //here we see what was removed
+  res.status(200).json({ task });
+
+  // res.status(200).send();
+  // res.status(200).json({ task: null, status: "success" });
+  //   } catch (error) {
+  //     res.status(500).json({ msg: error });
+  //   }
+});
+
+//PATCH - we update part ot item
+
+const updateTask = asyncWrapper(async (req, res) => {
```

Comment:
You need to add the next parameter here. When a task is not found, the app throws an error 500 because the next function is not defined.

Reply:
oh no! Thank you!